Changes in module 'nf-core/vardictjava'
--- modules/nf-core/vardictjava/main.nf
+++ modules/nf-core/vardictjava/main.nf
@@ -13,7 +13,7 @@
     tuple val(meta3), path(fasta_fai)
 
     output:
-    tuple val(meta), path("*.vcf"), emit: vcf
+    tuple val(meta), path("*.vcf.gz"), emit: vcf
     path "versions.yml"           , emit: versions
 
     when:
@@ -39,7 +39,7 @@
     | ${filter} \\
     | ${convert_to_vcf} \\
         ${args2} \\
-    > ${prefix}.vcf
+        | gzip -c > ${prefix}.vcf.gz
 
     cat <<-END_VERSIONS > versions.yml
     "${task.process}":

************************************************************
