Changes in module 'nf-core/gatk4/mutect2'
--- modules/nf-core/gatk4/mutect2/main.nf
+++ modules/nf-core/gatk4/mutect2/main.nf
@@ -14,8 +14,8 @@
     tuple val(meta4), path(dict)
     path(germline_resource)
     path(germline_resource_tbi)
-    path(panel_of_normals)
-    path(panel_of_normals_tbi)
+    // path(panel_of_normals)
+    // path(panel_of_normals_tbi)
 
     output:
     tuple val(meta), path("*.vcf.gz")     , emit: vcf
@@ -32,7 +32,7 @@
     def prefix = task.ext.prefix ?: "${meta.id}"
     def inputs = input.collect{ "--input $it"}.join(" ")
     def interval_command = intervals ? "--intervals $intervals" : ""
-    def pon_command = panel_of_normals ? "--panel-of-normals $panel_of_normals" : ""
+    // def pon_command = panel_of_normals ? "--panel-of-normals $panel_of_normals" : ""
     def gr_command = germline_resource ? "--germline-resource $germline_resource" : ""
 
     def avail_mem = 3072
@@ -47,7 +47,6 @@
         $inputs \\
         --output ${prefix}.vcf.gz \\
         --reference $fasta \\
-        $pon_command \\
         $gr_command \\
         $interval_command \\
         --tmp-dir . \\

************************************************************
